src/internal_modules/roc_core/heap_arena.cpp

```diff
@@ -62,12 +75,32 @@ void HeapArena::deallocate(void* ptr) {
         roc_panic("heap arena: unpaired deallocate");
     }
 
-    Chunk* chunk = ROC_CONTAINER_OF(ptr, Chunk, data);
+    Chunk* chunk = ROC_CONTAINER_OF((char*)ptr - sizeof(ChunkCanary), Chunk, data);
+
+    size_t size = chunk->size;
+    size_t size_padding = AlignOps::align_max(size) - size;
 
-    MemoryOps::poison_after_use(chunk->data, chunk->size);
+    char* canary_before = (char*)chunk->data;
+    char* memory = (char*)chunk->data + sizeof(ChunkCanary);
+    char* canary_after = (char*)chunk->data + sizeof(ChunkCanary) + size;
+
+    bool canary_before_ok = MemoryOps::check_canary(canary_before, sizeof(ChunkCanary));
+    bool canary_after_ok =
+        MemoryOps::check_canary(canary_after, size_padding + sizeof(ChunkCanary));
+    if (!canary_before_ok || !canary_after_ok) {
+        num_guard_failures_++;
+        if ((flags_ & HeapArenaFlag_EnableGuards) != 0) {
+            roc_panic("heap arena: buffer overflow detected");
+        }
```

Comment:
```suggestion
        if (flags_ & HeapArenaFlag_EnableGuards) {
            roc_panic("heap arena: detected memory violation: ok_before=%d ok_after=%d",
                      (int)canary_before_ok, (int)canary_after_ok);
        }
```

(same as here: https://github.com/roc-streaming/roc-toolkit/pull/594#issuecomment-1750704681)